### PR TITLE
Adds download_signed_apk_from_google_play lane

### PR DESCRIPTION
**Pull request by oguzkocer**
For now, this PR is only adding the standalone `download_signed_apk_from_google_play` lane, but I hope to follow it with another PR to attach the downloaded apks to the GitHub release.

Considering this lane will mostly be used during a release action, I wanted to keep the download location simple. So, I utilized the `bundle_file_path` and just replaced the `.aab` extension with `.apk`.

Thanks again @AliSoftware for adding this to `fastlane`. This has been the most frustrating part of the release process for me due to that awful Play Store UI, so I can't tell you how much I appreciate it! 🙇‍♂️ 

**To test:**
`bundle exec fastlane download_signed_apk_from_google_play app:jetpack`

## Regression Notes
1. Potential unintended areas of impact
N/A

2. What I did to test those areas of impact (or what existing automated tests I relied on)
N/A

3. What automated tests I added (or what prevented me from doing so)
N/A

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

UI Changes testing checklist:

- [x] Portrait and landscape orientations.
- [x] Light and dark modes.
- [x] Fonts: Larger, smaller and bold text.
- [x] High contrast.
- [x] Talkback.
- [x] Languages with large words or with letters/accents not frequently used in English.
- [x] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [x] Large and small screen sizes. (Tablet and smaller phones)
- [x] Multi-tasking: Split screen and Pop-up view. (Android 10 or higher)

### CODE REVIEW

File: fastlane/lanes/build.rb
Comment (AliSoftware) [inline fastlane/lanes/build.rb:143]: I don't see from the code how passing `version:12.3,789` would allow you to provide the version name and code in one comma-separated parameter like how it's suggested in this doc?

File: fastlane/lanes/build.rb
Comment (AliSoftware) [inline fastlane/lanes/build.rb]: iinm this line would make `version` be a `String` if passed explicitly via `version:xyz` option, while it would be a `Hash` if not passed as an option and we end up calling `android_get_release_version`, no?

Which would make the rest of the code trying to read things like `version[:code]` a couple of lines below crash if provided as an option aka a `String` not a `Hash`?

File: fastlane/lanes/release.rb
Comment (AliSoftware) [inline fastlane/lanes/release.rb]: ~If I'm reading this correctly, the call site in your `download_signed_apk_from_google_play` lane calls this with `version` being a `Hash`, not a `String` like this code seems to expect?~ [EDIT] nvm i now see that the logic to untangle the `version` Hash is handled by `bundle_file_path` itself. [/EDIT]

Also, I'm surprised that using `sub!`, aka the mutating-in-place version of `sub` (that's what the `!` in the method name means), on the return value of a function call does not crash. I'm pretty sure in strict mode it would? And there's no real need to mutate in place here anyway?

**Review by AliSoftware — COMMENTED** [review]
Not on my Mac so haven't tested the command in a terminal, so those are just review from reading the code without testing.

File: fastlane/lanes/build.rb
Comment (oguzkocer) [reply fastlane/lanes/build.rb:143]: I actually tried to figure out how we were supposed to pass a Hash from command line and couldn't do it. So, I copied this documentation format from `build_bundle` lane. I figured, maybe this is how we are documenting it for when the value is passed from another lane.

I'd appreciate it if you can share the appropriate way to handle this. Thanks!

File: fastlane/lanes/build.rb
Comment (oguzkocer) [reply fastlane/lanes/build.rb]: I think [my previous comment](https://github.com/wordpress-mobile/WordPress-Android/pull/18813#discussion_r1271162781) about the documentation covers this. This is just how the version is handled across other lanes, so I am just following our current practice.

File: fastlane/lanes/release.rb
Comment (oguzkocer) [reply fastlane/lanes/release.rb]: Addressed in 5727395955adc4cf38ebe77c4a217c727eea0370.

File: fastlane/lanes/build.rb
Comment (AliSoftware) [reply fastlane/lanes/build.rb:143]: Yeah unfortunately, to my knowledge at least, there's no easy way to provide a Hash value via command line.

From what I see, the way this has been handled in other lanes is by explicitly testing if `version` is a `String` after pulling its value from `options[:version]`, and if so then split the string with `,` and reconstruct the `Hash`:
https://github.com/wordpress-mobile/WordPress-Android/blob/bb4eb2d39f6ff2bf95f53c337bf2bb12a03f9f72/fastlane/lanes/build.rb#L203-L206

---

The alternative solution would be the other way around, i.e. have the lane take the options separately (`options[:versionName]` and `options[:versionCode]`), so you can provide separate String values from the command line; and then, for when you call the lane via another lane which has the `version` in a variable as a `Hash` (from calling `version = android_get_release_version()`), we'd have to explicitly pass `versionName: version['name'], versionCode: version['code']` as separate parameters…

Up to you. But given other lanes are using the `if version.is_a?(String) … ` trick so far, I guess you it would make sense to continue that way for your new lane here?

Comment (oguzkocer) [reply fastlane/lanes/build.rb:143]: Thanks a lot for all this explanation @AliSoftware! I should have checked if the `version` was handled in a different way for these lanes, but I am really not used to this kind of pattern - as they are impossible in typed languages - so I didn't expect it at all.

I've made the changes in dbe95dfb2305334fa4aed88034554ea8b23d5223 and ran some random tests that seems to work as expected:

```
bundle exec fastlane download_signed_apk_from_google_play app:wordpress version:22.9-rc-1,1359
bundle exec fastlane download_signed_apk_from_google_play app:wordpress version:22-rc-1,1359
bundle exec fastlane download_signed_apk_from_google_play app:jetpack version:arst,1300
```

Comment (AliSoftware) [reply fastlane/lanes/build.rb:143]: > but I am really not used to this kind of pattern - as they are impossible in typed languages

Oh yeah I'm with you on that, this is not super pretty. Which is why I also mentioned the second option / alternative solution to pass those as separate options to the lane — so command line calls can pass them separately, while parent lanes calling that lane can pass each entry of the Hash separately — which I feel would be much cleaner to understand and would avoid us to rely on having to test the type of the variable with `is_a?(String)`!

So feel free to adopt this alternative solution at some point in a subsequent PR, to clean up the logic and implementation if you want to make this less confusing / convoluted / less relying on Ruby duck-typing.

File: fastlane/lanes/build.rb
Comment (AliSoftware) [inline fastlane/lanes/build.rb]: In practice, for this specific use case, we probably don't want to fallback to using a `versionCode` of `1` because the GooglePlay API will really require the real versionCode to know what APK to download.
```suggestion
      UI.user_error!('Please pass the `version` option as a comma-separated `name,code` value') if version_code.nil?
      version = { 'name' => version_name, 'code' => version_code }
```

**Review by AliSoftware — APPROVED** [review]
Disabled auto-merge in case you want to address the small note I left about default value for `version_code` before merging, but otherwise looks good to me, so approving to unblock 👍 

PS: `Gemfile.lock` conflict is likely due to https://github.com/wordpress-mobile/WordPress-Android/pull/18848 landing first 🙃

File: fastlane/lanes/build.rb
Comment (AliSoftware) [reply fastlane/lanes/build.rb:143]: TIL: after randomly browsing the source code for fastlane's `ConfigItem` class, I just discovered [that fastlane _does_ support passing a Hash from the command line… as long as you use valid JSON syntax](https://github.com/fastlane/fastlane/blob/master/fastlane_core/lib/fastlane_core/configuration/config_item.rb#L283-L288) 🤯   Didn't know that.

That being said, this would make the command line call quite ugly to write for that case… so it's probably not the direction we want to go to allow this lane to be called from CLI 🙃

File: fastlane/lanes/build.rb
Comment (oguzkocer) [reply fastlane/lanes/build.rb]: @AliSoftware I have been downloading the APK by running `bundle exec fastlane download_signed_apk_from_google_play app:jetpack` which uses the version name & version code for the current branch. This has been working very well for me.

Could you elaborate on your concerns about having a default value? I feel like it's a valid case, but maybe I am missing something?

Comment (oguzkocer) [reply fastlane/lanes/build.rb]: Oh, this is in a completely different place. Sorry @AliSoftware - I shouldn't work before I put my lenses on. I'll take a look at this later today. 🤦

Comment (AliSoftware) [reply fastlane/lanes/build.rb]: My note was about if you _do_ provide a `version:` from the command line… but without the expected comma-separate values format. Like `bundle exec fastlane download_signed_apk_from_google_play app:jetpack version:22.9-rc-1`.

In such cases, the branch with `is_a?(String)` would be taken for the version value provided by command line, and the  `split` call would extract `22.9-rc-1` into `version_name`, but leave `version_code` as `nil` because you didn't provide a comma + version code in command line option. In such case, that default value of `1` for `version_code` would make it behave like if you provided `version:22.9-rc-1,1` to the command line (instead of `22.9-rc-1,1359`)

---

To avoid that case, my code suggestion above would just error if you accidentally tried to provide a `version:` parameter from the command line but without a comma.

Alternatively you could consider reading the `versionCode` from the current version file (while having the versionName be overriden by the value passed as `version:22.9-rc-1` parameter)… but this doesn't seem worth it to cover those twisted cases — especially when I don't expect you to provide the `version:` option via the command line in 99% of use cases of this lane anyway. And also, all those edge cases when providing the value from command line would go away if you decide to refactor the lane options in a subsequent PR at some point to pass those as two separate parameters to make the whole thing less convoluted, [as discussed here](https://github.com/wordpress-mobile/WordPress-Android/pull/18813#discussion_r1277360443)… so probably not worth spending more time on it than just the basic `UI.user_error!` protection I suggested above to avoid accidental misuse?

Comment (oguzkocer) [reply fastlane/lanes/build.rb]: @AliSoftware You probably missed [my last comment](https://github.com/wordpress-mobile/WordPress-Android/pull/18813#discussion_r1277764744), but I deserve that wall of text. Sorry about wasting your time 😢

Comment (AliSoftware) [reply fastlane/lanes/build.rb]: No worries 🙂  🤗  ☕  👓

File: fastlane/lanes/build.rb
Comment (oguzkocer) [reply fastlane/lanes/build.rb:143]: I could swear that I've seen it before _somewhere_, but it might still be a delusion. Anyhow, thank you for finding this!

File: fastlane/lanes/build.rb
Comment (oguzkocer) [reply fastlane/lanes/build.rb]: Now that I can actually see what the suggestion is, this makes a lot of sense. Thank you for the suggestion and your understanding about the earlier mistake @AliSoftware 🙇‍♂️ 

P.S: This has not been a good morning so far, just after posting that comment, I also accidentally merged `trunk` into this branch instead of the `release/22.9` which is why there is a force push about an hour ago. Hopefully rest of the day will be better 🤞

Comment (AliSoftware) [reply fastlane/lanes/build.rb]: virtual-hug

**Pull request merged**
